File: src/hooks/stop.ts
import { extractCwd, isConfigured, loadHookConfig, postMemoryEvent, readHookInput, writeHookOutput } from "./common.js";

export async function runStopHook(): Promise<void> {
  if (process.env.CONTEXTSTREAM_STOP_ENABLED === "false") {
    writeHookOutput();
    return;
  }

  const input = readHookInput<Record<string, unknown>>();
  const cwd = extractCwd(input);
  const config = loadHookConfig(cwd);

  if (isConfigured(config)) {
    const sessionId = (typeof input.session_id === "string" && input.session_id) || "unknown";
    const reason =
      (typeof input.reason === "string" && input.reason) ||
      (typeof input.stop_reason === "string" && input.stop_reason) ||
      "response_complete";

    await postMemoryEvent(
      config,
      "Stop checkpoint",
      {
        session_id: sessionId,
        reason,
        hook: "stop",
        timestamp: new Date().toISOString(),
        tool_name: typeof input.tool_name === "string" ? input.tool_name : null,
        model: typeof input.model === "string" ? input.model : null,
      },
      ["hook", "stop", "checkpoint"]
    ).catch(() => {});
  }

  writeHookOutput();
}

const isDirectRun = process.argv[1]?.includes("stop") || process.argv[2] === "stop";
if (isDirectRun) {
  runStopHook().catch(() => process.exit(0));
}

File: crates/mcp-server/src/hook_handlers/stop.rs
//! Stop hook handler.
//!
//! Fires whenever Claude finishes responding. Captures a lightweight checkpoint
//! so recent progress is not lost between full SessionEnd events.

use anyhow::Result;

use super::common::{extract_cwd, load_config, post_memory_event};
use super::{read_stdin_json, write_stdout_json, HookOutput};

/// Handle the Stop hook.
pub async fn handle() -> Result<()> {
    if std::env::var("CONTEXTSTREAM_STOP_ENABLED")
        .map(|v| v == "false")
        .unwrap_or(false)
    {
        write_stdout_json(&HookOutput::empty())?;
        return Ok(());
    }

    let input = read_stdin_json()?;
    let cwd = extract_cwd(&input);
    let config = load_config(&cwd);

    if config.is_configured() {
        let session_id = input
            .get("session_id")
            .and_then(|v| v.as_str())
            .unwrap_or("unknown");
        let reason = input
            .get("reason")
            .or_else(|| input.get("stop_reason"))
            .and_then(|v| v.as_str())
            .unwrap_or("response_complete");
        let last_assistant_message = input
            .get("last_assistant_message")
            .and_then(|v| v.as_str())
            .map(str::trim)
            .filter(|value| !value.is_empty());

        let content = serde_json::json!({
            "session_id": session_id,
            "reason": reason,
            "hook": "stop",
            "timestamp": chrono::Utc::now().to_rfc3339(),
            "tool_name": input.get("tool_name").and_then(|v| v.as_str()),
            "model": input.get("model").and_then(|v| v.as_str()),
            "stop_hook_active": input.get("stop_hook_active").and_then(|v| v.as_bool()),
            "last_assistant_message": last_assistant_message,
        });

        post_memory_event(
            &config,
            "Stop checkpoint",
            content,
            &["hook", "stop", "checkpoint"],
        )
        .await;
    }

    // Stop hooks are side-effect only.
    write_stdout_json(&HookOutput::empty())?;
    Ok(())
}
